shared/src/main/scala/eu/timepit/refined/api/Validate.scala

```diff
@@ -0,0 +1,113 @@
+package eu.timepit.refined
+package api
+
+import eu.timepit.refined.internal.Resources
+
+import scala.util.Try
+
+/**
+ * Type class for validating values of type `T` according to a type-level
+ * predicate `P`. The semantics of `P` are defined by the instance(s) of
+ * this type class for `P`.
+ */
+trait Validate[T, P] extends Serializable { self =>
+
+  type R
+
+  final type Res = Result[R]
+
+  def validate(t: T): Res
+
+  /** Returns a string representation of this [[Validate]] using `t`. */
+  def showExpr(t: T): String
+
+  def showResult(t: T, r: Res): String =
+    Resources.predicateResultDetailDot(r, showExpr(t))
+
+  /**
+   * Denotes whether this [[Validate]] is constant (which is false by
+   * default). A constant [[Validate]] ignores the argument passed to
+   * [[validate]].
+   */
+  val isConstant: Boolean = false
```

Comment:
The reason is outlined in #26. See also https://github.com/fthomas/refined/blob/master/docs/ctorfieldnames.md

I admit that this might not look very compelling but it is kind of cool that the `ConstructorNames` and `FieldNames` predicates can be evaluated at compile-time. :) But this is also present in the current `Predicate` and nothing new that has been introduced in this PR.